pkg/pdp/aggregator/fns/fns.go

```diff
@@ -0,0 +1,147 @@
+package fns
+
+import (
+	"cmp"
+	"context"
+
+	// for go:embed
+	_ "embed"
+	"fmt"
+	"slices"
+
+	ipldprime "github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/schema"
+	"github.com/storacha/go-capabilities/pkg/pdp"
+	"github.com/storacha/go-piece/pkg/piece"
+	"github.com/storacha/go-ucanto/core/invocation/ran"
+	"github.com/storacha/go-ucanto/core/ipld"
+	"github.com/storacha/go-ucanto/core/iterable"
+	"github.com/storacha/go-ucanto/core/receipt"
+	"github.com/storacha/go-ucanto/core/result"
+	"github.com/storacha/go-ucanto/ucan"
+	"github.com/storacha/storage/pkg/pdp/aggregator/aggregate"
+	"github.com/storacha/storage/pkg/pdp/curio"
+)
+
+//go:embed buffer.ipldsch
+var bufferSchema []byte
+
+var bufferTS *schema.TypeSystem
+
+func init() {
+	ts, err := ipldprime.LoadSchemaBytes(bufferSchema)
+	if err != nil {
+		panic(fmt.Errorf("loading blob schema: %w", err))
+	}
+	bufferTS = ts
+}
+
+func BufferType() schema.Type {
+	return bufferTS.TypeByName("Buffer")
+}
+
+// Buffer tracks in progress work building an aggregation
+type Buffer struct {
+	TotalSize           uint64
+	ReverseSortedPieces []piece.PieceLink
+}
+
+// InsertReverseSortedBySize adds a piece to a list of pieces sorted largest to smallest, maintaining sort order
+func InsertReverseSortedBySize(sortedPieces []piece.PieceLink, newPiece piece.PieceLink) []piece.PieceLink {
+	pos, _ := slices.BinarySearchFunc(sortedPieces, newPiece, func(test, target piece.PieceLink) int {
+		// flip ordering comparing size cause we're going in reverse order
+		return cmp.Compare(target.PaddedSize(), test.PaddedSize())
+	})
+	return slices.Insert(sortedPieces, pos, newPiece)
+}
+
+// MinAggregateSize is 128MB
+// Max size is 256MB -- this means we will never see an individual piece larger
+// than 256MB -- the upload will fail otherwise
+// So we can safely assume that if we see a 256MB piece, we just submit immediately
+// If not, we can safely aggregate till >=128MB without going over 256MB
+const MinAggregateSize = 128 << 20
+
+func AggregatePiece(buffer Buffer, newPiece piece.PieceLink) (Buffer, *aggregate.Aggregate, error) {
+	// if the piece is aggregatable on its own it should submit immediately
+	if newPiece.PaddedSize() > MinAggregateSize {
+		aggregate, err := aggregate.NewAggregate([]piece.PieceLink{newPiece})
+		return buffer, &aggregate, err
+	}
+
+	newSize := buffer.TotalSize + newPiece.PaddedSize()
+	newPieces := InsertReverseSortedBySize(buffer.ReverseSortedPieces, newPiece)
+
+	// if we have reached the minimum aggregate size, submit and start over
+	if newSize >= MinAggregateSize {
+		aggregate, err := aggregate.NewAggregate(newPieces)
+		return Buffer{}, &aggregate, err
```

Comment:
If there was an error here, shouldn't we return the buffer untouched? Otherwise we could lose pieces.

```suggestion
		if err != nil {
			return buffer, nil, err
		}
		return Buffer{}, &aggregate, err
```